### PR TITLE
Check if subpath can do motion.

### PR DESCRIPTION
**Pull request by InnerLife0**
`cdbpath_create_motion_path` function doesn't check subpath's params Relids. When Path contains index scan with quals it leads to segfault.
Fix adds a few additional checks of can motion node be generated atop of subpath. Additional regression test included.
Related to issue #[12228](https://github.com/greenplum-db/gpdb/issues/12228). Partial backport of 9cc1da6 (#[10012](https://github.com/greenplum-db/gpdb/issues/10012)).

### CODE REVIEW

**Comment by InnerLife0** [comment]
@kainwen I require your approval, because you are already into this.

**Review by kainwen — APPROVED** [review]
LGTM.

Can we mention the github issue as comment in the test case? Thanks!

It seems pipeline is not triggered, I have built a full dev pipeline to run it.

**Comment by kainwen** [comment]
@InnerLife0 The only failure case is (under planner), would you please help fix this in your PR? Thanks.

```diff
--- \/tmp\/build\/e18b2f02\/gpdb_src\/src\/test\/regress\/expected\/gporca\.out	2021-06-30 22:37:51.416598687 +0000
+++ \/tmp\/build\/e18b2f02\/gpdb_src\/src\/test\/regress\/results\/gporca\.out	2021-06-30 22:37:51.892639945 +0000
@@ -13955,28 +13991,7 @@
 from tcorr1 out
 where out.b in (select coalesce(tcorr2.a, 99)
                 from tcorr1 left outer join tcorr2 on tcorr1.a=tcorr2.a+out.a);
-                                                         QUERY PLAN                                                         
-----------------------------------------------------------------------------------------------------------------------------
-                                             ->  Seq Scan on tcorr1  (cost=0.00..1.01 rows=1 width=4)
-                                             ->  Seq Scan on tcorr2  (cost=0.00..1.01 rows=1 width=4)
-                                             Hash Key: (tcorr2.a + "out".a)
-                                       ->  Hash  (cost=1.01..1.01 rows=1 width=4)
-                                       ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.03 rows=1 width=4)
-                                       Hash Cond: ((tcorr2.a + "out".a) = tcorr1.a)
-                                 ->  Hash Right Join  (cost=1.02..2.09 rows=2 width=4)
-                                 Filter: ("out".b = "ANY_subquery"."coalesce")
-                           ->  Seq Scan on tcorr1 "out"  (cost=0.00..1.01 rows=1 width=18)
-                           ->  Subquery Scan on "ANY_subquery"  (cost=1.02..2.13 rows=1 width=4)
-                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..1.05 rows=1 width=18)
-                     ->  Materialize  (cost=1.02..2.14 rows=1 width=4)
-               ->  Nested Loop  (cost=10000000001.02..10000000003.23 rows=2 width=18)
-               Hash Key: "out".ctid
-         ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=10000000001.02..10000000003.30 rows=2 width=18)
-         Group Key: "out".ctid, "out".gp_segment_id
-   ->  HashAggregate  (cost=10000000003.32..10000000003.35 rows=2 width=18)
- Gather Motion 3:1  (slice4; segments: 3)  (cost=10000000003.32..10000000003.35 rows=4 width=18)
-(19 rows)
-
+ERROR:  Unexpected internal error (pathnode.c:1887)
 -- expect 1 row
 -- FIXME: A process terminates during execution, see https://github.com/greenplum-db/gpdb/issues/10791
 -- select *
@@ -14044,28 +14059,7 @@
 from tcorr1 out
 where out.b in (select coalesce(tcorr2.a, 99)
                 from tcorr1 left outer join tcorr2 on tcorr1.a=tcorr2.a+out.a);
-                                                         QUERY PLAN                                                         
-----------------------------------------------------------------------------------------------------------------------------
-                                             ->  Seq Scan on tcorr1  (cost=0.00..1.01 rows=1 width=4)
-                                             ->  Seq Scan on tcorr2  (cost=0.00..1.01 rows=1 width=4)
-                                             Hash Key: (tcorr2.a + "out".a)
-                                       ->  Hash  (cost=1.01..1.01 rows=1 width=4)
-                                       ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.03 rows=1 width=4)
-                                       Hash Cond: ((tcorr2.a + "out".a) = tcorr1.a)
-                                 ->  Hash Right Join  (cost=1.02..2.09 rows=2 width=4)
-                                 Filter: ("out".b = "ANY_subquery"."coalesce")
-                           ->  Seq Scan on tcorr1 "out"  (cost=0.00..1.01 rows=1 width=18)
-                           ->  Subquery Scan on "ANY_subquery"  (cost=1.02..2.13 rows=1 width=4)
-                     ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..1.05 rows=1 width=18)
-                     ->  Materialize  (cost=1.02..2.14 rows=1 width=4)
-               ->  Nested Loop  (cost=10000000001.02..10000000003.23 rows=2 width=18)
-               Hash Key: "out".ctid
-         ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=10000000001.02..10000000003.30 rows=2 width=18)
-         Group Key: "out".ctid, "out".gp_segment_id
-   ->  HashAggregate  (cost=10000000003.32..10000000003.35 rows=2 width=18)
- Gather Motion 3:1  (slice4; segments: 3)  (cost=10000000003.32..10000000003.35 rows=4 width=18)
-(19 rows)
-
+ERROR:  Unexpected internal error (pathnode.c:1887)
 -- expect 1 row
 -- FIXME: A process terminates during execution, see https://github.com/greenplum-db/gpdb/issues/10791
 -- select *

======================================================================
```

**Comment by InnerLife0** [comment]
> Can we mention the github issue as comment in the test case? Thanks!

@kainwen Done.

> The only failure case is (under planner), would you please help fix this in your PR? Thanks.

For some reason I was not able to reproduce this error running `installcheck-world` locally, but was able to reproduce it manually.
I removed default check at the end of `cdbpath_create_motion_path` function, but we should think about correcting this test somewhere in another PR. This test produces a plan which can't be executed (this is why next test is commented out).

File: src/backend/cdb/cdbpath.c [hunk 1]
```diff
@@ -304,6 +314,11 @@ cdbpath_create_motion_path(PlannerInfo *root,
 	/* Does subpath produce same multiset of rows on every qExec of its gang? */
 	else if (CdbPathLocus_IsReplicated(subpath->locus))
 	{
+		/*
+		 * If the subpath requires parameters, we cannot generate Motion atop of it.
+		 */
+		if (!bms_is_empty(PATH_REQ_OUTER(subpath)))
+			return NULL;
```
Comment (maksm90) [inline src/backend/cdb/cdbpath.c:321]: The replicated locus type case for subpath doesn't produce motion upper node - it's clear on code below therefore the check on request of outer parameters here is needless.

File: src/backend/cdb/cdbpath.c [hunk 2]
```diff
@@ -139,6 +139,11 @@ cdbpath_create_motion_path(PlannerInfo *root,
 		/* singleQE-->entry?  Don't move.  Slice's QE will run on entry db. */
 		if (CdbPathLocus_IsSingleQE(subpath->locus))
 		{
+			/*
+			 * If the subpath requires parameters, we cannot generate Motion atop of it.
+			 */
+			if (!bms_is_empty(PATH_REQ_OUTER(subpath)))
+				return NULL;
```
Comment (maksm90) [inline src/backend/cdb/cdbpath.c:146]: Please, add blank line after block before comment.

File: src/backend/cdb/cdbpath.c [hunk 3]
```diff
@@ -170,6 +175,11 @@ cdbpath_create_motion_path(PlannerInfo *root,
 
 		if (CdbPathLocus_IsSegmentGeneral(subpath->locus))
 		{
+			/*
+			 * If the subpath requires parameters, we cannot generate Motion atop of it.
+			 */
+			if (!bms_is_empty(PATH_REQ_OUTER(subpath)))
+				return NULL;
```
Comment (maksm90) [inline src/backend/cdb/cdbpath.c:182]: Similarly, blank line below.

**Review by maksm90 — COMMENTED** [review]
As you might mention in original patch 9cc1da6 all checks on request of outer params for subpath precede all possible cases of generating motion path. Your patch doesn't inject this check before the [default case](https://github.com/arenadata/gpdb/blob/ADBDEV-1715-2/src/backend/cdb/cdbpath.c#L373-L404) after all partial cases. Original patch [does](https://github.com/arenadata/gpdb/commit/9cc1da6163d894d638e2455e046dfdbf0763725b#diff-279922ddd4ce14f29db00d31513dc73092c3576b0fe8d73ae6fd67a364124eb1R498-R503) this.

File: src/backend/cdb/cdbpath.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/cdb/cdbpath.c:321]: Fixed in #12287.

File: src/backend/cdb/cdbpath.c (hunk 2 above)
Comment (InnerLife0) [reply src/backend/cdb/cdbpath.c:146]: Fixed in #12287.

File: src/backend/cdb/cdbpath.c (hunk 3 above)
Comment (InnerLife0) [reply src/backend/cdb/cdbpath.c:182]: Fixed in #12287.

**Comment by InnerLife0** [comment]
> As you might mention in original patch [9cc1da6](https://github.com/greenplum-db/gpdb/commit/9cc1da6163d894d638e2455e046dfdbf0763725b) all checks on request of outer params for subpath precede all possible cases of generating motion path. Your patch doesn't inject this check before the [default case](https://github.com/arenadata/gpdb/blob/ADBDEV-1715-2/src/backend/cdb/cdbpath.c#L373-L404) after all partial cases. Original patch [does](https://github.com/arenadata/gpdb/commit/9cc1da6163d894d638e2455e046dfdbf0763725b#diff-279922ddd4ce14f29db00d31513dc73092c3576b0fe8d73ae6fd67a364124eb1R498-R503) this.

Added TODO in #12287, as we can't add check to default case for [now](https://github.com/greenplum-db/gpdb/pull/12238#issuecomment-871973153).

**Pull request merged**